thunder/clang/__init__.py

```diff
@@ -776,17 +829,42 @@ def _advanced_indexing(a: TensorLike, /, key) -> TensorLike:
     #   but actually generating those tensors seems expensive.
     flattened: TensorLike
     subtensor_shape: list[int]
+    # keeps track of the rest of the dimensions
+    remaining_shape: list[int]
     dim: int
     modified_key = list(key)
-
+    # calculates the output tensor shape
+    new_shape = []
+
+    permute_shape = []  # basic keys first followed by advanced keys
+    permute_key = []
+    for key_idx, key_val in basic_keys:
+        permute_shape.append(key_idx)
+        permute_key.append(key_val)
+    for key_idx, key_val in advanced_keys:
+        permute_shape.append(key_idx)
+        permute_key.append(key_val)
+
+    # transpose so that we have basic keys first
     if has_ellipsis:
         del modified_key[0]
+        permute_shape = [x + len(list(a.shape[: a.ndim - (seq_len - 1)])) - 1 for x in permute_shape]
+        permute_shape = list(range(a.ndim - (seq_len - 1))) + permute_shape
+        a = transpose(a, tuple(permute_shape))
+        key = tuple(permute_key)
+
         subtensor_shape = a.shape[a.ndim - (seq_len - 1) :]
+        remaining_shape = list(a.shape[: a.ndim - (seq_len - 1)])
         dim = -1
         flattened = flatten(a, a.ndim - (seq_len - 1), -1)
     else:
         # NOTE No ellipsis case
+        permute_shape = permute_shape + list(range(len(key), len(a.shape)))
+        a = transpose(a, tuple(permute_shape))
```

Comment:
Same as above (Should this be with an if in case we already have the right permutation?).